src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java

```diff
@@ -346,19 +349,71 @@ private void createAssemblyArchive(AssemblyConfiguration assemblyConfig, MojoPar
+            String error = "Failed to create assembly for docker image " +
+                           " (with mode '" + buildMode + "'): " + e.getMessage() + ".";
+            if (params.getProject().getArtifact().getFile() == null) {
+                error += " If you include the build artifact please ensure that you have " +
+                         "built the artifact before with 'mvn package' (should be available in the target/ dir). " +
+                         "Please see the documentation (section \"Assembly\") for more information.";
+            }
+            throw new MojoExecutionException(error, e);
         } catch (InvalidAssemblerConfigurationException e) {
             throw new MojoExecutionException(assembly, "Assembly is incorrectly configured: " + assembly.getId(),
                                             "Assembly: " + assembly.getId() + " is not configured correctly: "
                                             + e.getMessage());
+        } finally {
+            setArtifactFile(params.getProject(), originalArtifactFile);
+        }
+    }
+
+    // Set an artifact file if it is missing. This workaround the issues
+    // mentioned first in https://issues.apache.org/jira/browse/MASSEMBLY-94 which requires the package
+    // phase to run so set the ArtifactFile. There is no good solution, so we are trying
+    // to be very defensive and add a workaround for some situation which won't work for every occasion.
+    // Unfortunately a plain forking of the Maven lifecycle is not good enough, since the MavenProject
+    // gets cloned before the fork, and the 'package' plugin (e.g. JarPlugin) sets the file on the cloned
+    // object which is then not available for the BuildMojo (there the file is still null leading to the
+    // the "Cannot include project artifact: ... The following patterns were never triggered in this artifact inclusion filter: <artifact>"
+    // warning with an error following.
+    private File ensureThatArtifactFileIsSet(MavenProject project) {
+        Artifact artifact = project.getArtifact();
+        if (artifact == null) {
+            return null;
+        }
+        File oldFile = artifact.getFile();
+        if (oldFile != null) {
+            return oldFile;
         }
+        Build build = project.getBuild();
+        if (build == null) {
+            return null;
+        }
+        String finalName = build.getFinalName();
+        String target = build.getDirectory();
+        if (finalName == null || target == null) {
+            return null;
+        }
+        File artifactFile = new File(target, finalName + "." + project.getPackaging());
+        if (artifactFile.exists() && artifactFile.isFile()) {
+            // TODO: Maybe check also that the artifact file is 'young' ? (so that it could be assumed that
+            //       it has been created recently by a 'package' in the same run.)
+            setArtifactFile(project, artifactFile);
```

Comment:
actual not, because the return value is used as the old value, to reset the file again. Since its was already null before, null is the proper return value.